File: mooreTracing.m
function [B,L] = mooreTracing(region)
B = {};
L = {};

traceRegion = region; % changing the binary region by placing a -1 one in place of the border starting value
marker = zeros(size(traceRegion)); %Use this adjacent matrix / overlapping matrix

%disp(traceRegion(590:630,1970:2000));

[x,y] = size(region);
numcells = 0; % Current cell of B

for i = 1:x,
    for j = 1:y,

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        if(and(j + 1 < y,i + 1 < x)),
            entryLocation = []; 
            s = []; 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            if((traceRegion(i,j+1) == 0) && traceRegion(i,j) ~= traceRegion(i,j+1) && marker(i,j+1)~=-1), %white to black detection
                numcells = numcells + 1;
                numRows = 1;
                prev = [];
                entryLocation(1) = i;
                entryLocation(2) = j;
                s(1) = i;
                s(2) = j+1;
                prev(1) = i;
                prev(2) = j;
                currentPx = s(1);
                currentPy = s(2);
                B{numcells}(numRows,1) = currentPx;
                B{numcells}(numRows,2) = currentPy;
                marker(currentPx,currentPy) = -1;
                
                numRows = numRows + 1;
                [currentPx,currentPy,prev] = getNeighbourhood(traceRegion,currentPx,currentPy,prev);
                B{numcells}(numRows,1) = currentPx;
                B{numcells}(numRows,2) = currentPy;
                marker(currentPx,currentPy) = -1;
                
                while (s(1) ~= currentPx) | (s(2)~=currentPy),
                    numRows = numRows + 1;
                    [currentPx,currentPy,prev] = getNeighbourhood(traceRegion,currentPx,currentPy,prev);
                    B{numcells}(numRows,1) = currentPx;
                    B{numcells}(numRows,2) = currentPy;
                    marker(currentPx,currentPy) = -1;
                end
%                 disp(B{2});
%                 imshow(region);
%                 hold on
%                 boundary = B{8};
%                 plot(boundary(:,2), boundary(:,1), 'r', 'LineWidth', 2) 
            end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        end
    end
end

end

File: python/domino_helper/moore_tracing.py
"""
mooreTracing.m -> moore_tracing.py

Custom Moore neighborhood boundary tracing algorithm.
Scans the binary image for white-to-black transitions and traces each
boundary contour using the getNeighbourhood function.
"""

import numpy as np
from .get_neighbourhood import get_neighbourhood


def moore_tracing(region):
    """
    Trace boundaries in a binary image using Moore neighborhood tracing.

    Parameters
    ----------
    region : np.ndarray
        Binary image (values 0 and 1).

    Returns
    -------
    B : list of np.ndarray
        List of boundary arrays, each of shape (N, 2) with [row, col] coordinates.
    L : list
        Placeholder (matches MATLAB signature).
    """
    B = []
    L = []

    trace_region = region.copy()
    x, y = region.shape
    marker = np.zeros((x, y), dtype=np.int32)

    for i in range(x):
        for j in range(y):
            if j + 1 < y and i + 1 < x:
                # White to black detection
                if (trace_region[i, j + 1] == 0 and
                        trace_region[i, j] != trace_region[i, j + 1] and
                        marker[i, j + 1] != -1):

                    boundary = []

                    s = [i, j + 1]
                    prev = [i, j]
                    current_px = s[0]
                    current_py = s[1]

                    boundary.append([current_px, current_py])
                    marker[current_px, current_py] = -1

                    current_px, current_py, prev = get_neighbourhood(
                        trace_region, current_px, current_py, prev
                    )
                    boundary.append([current_px, current_py])
                    marker[current_px, current_py] = -1

                    while current_px != s[0] or current_py != s[1]:
                        current_px, current_py, prev = get_neighbourhood(
                            trace_region, current_px, current_py, prev
                        )
                        boundary.append([current_px, current_py])
                        marker[current_px, current_py] = -1

                    B.append(np.array(boundary))

    return B, L
